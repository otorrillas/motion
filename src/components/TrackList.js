import React from 'react';

import Track from './Track';

import styles from './TrackList.module.css';

const TrackList = ({ songs, onSelect }) => (
  <table className={styles.trackList}>
    <thead>
      <tr>
        <th />
        <th>Song</th>
        <th>Artist</th>
        <th>Album</th>
      </tr>
    </thead>
    <tbody>
      {songs &&
        songs.map(track => (
          <Track key={track.id} track={track} onClick={() => onSelect(track)} />
        ))}
    </tbody>
  </table>
);

export default TrackList;
