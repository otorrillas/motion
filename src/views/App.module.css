.app {
  color: white;
  text-align: center;

  padding-top: 4rem;
  /* display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center; */
}
