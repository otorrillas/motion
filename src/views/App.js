import React, { PureComponent, Suspense } from 'react';

import styles from './App.module.css';

import { fetchTrackList } from '../api';
import { PlayerContext } from './PlayerContext';

import Loader from '../components/Loader';
import TrackList from '../components/TrackList';
import Player from '../components/Player';

class App extends PureComponent {
  state = {
    trackList: null,
    currentTrack: null,
    isPlaying: false,
    playerRef: React.createRef()
  };

  componentDidMount() {
    this._loadAsyncData();
  }

  componentWillUnmount() {
    if (this._asyncRequest) {
      this._asyncRequest.cancel();
    }
  }

  _loadAsyncData() {
    this._asyncRequest = fetchTrackList().then(trackList => {
      this._asyncRequest = null;
      this.setState({ trackList });
    });
  }

  onTrackSelect = track => {
    const { currentTrack, isPlaying } = this.state;

    if (currentTrack && track === currentTrack) {
      const action = isPlaying ? this.pause : this.play;
      action();
    } else {
      this.setState({
        currentTrack: track,
        isPlaying: true
      });
    }
  };

  pause = () => {
    this.state.playerRef.current.pause();
    this.setState({
      isPlaying: false
    });
  };

  play = () => {
    this.state.playerRef.current.play();
    this.setState({
      isPlaying: true
    });
  };

  render() {
    const { currentTrack, trackList, playerRef } = this.state;

    return (
      <Suspense fallback={<Loader />} maxDuration={2000}>
        <div className={styles.app}>
          <h1>Motion</h1>

          <PlayerContext.Provider value={this.state}>
            <TrackList songs={trackList} onSelect={this.onTrackSelect} />

            <Player track={currentTrack} ref={playerRef} />
          </PlayerContext.Provider>
        </div>
      </Suspense>
    );
  }
}

export default App;
