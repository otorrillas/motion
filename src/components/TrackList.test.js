import React from 'react';
import renderer from 'react-test-renderer';

import TrackList from './TrackList';
import { PlayerContext } from '../views/PlayerContext';

describe('<TrackList />', () => {
  const mockActions = {
    onSelect: jest.fn()
  };

  const requiredProps = {
    songs: [
      {
        id: 123,
        title: 'sample title',
        artist: 'sample artist',
        album: 'sample album'
      },
      {
        id: 456,
        title: 'sample title 2',
        artist: 'sample artist 2',
        album: 'sample album 2'
      }
    ]
  };
  const renderComponent = (
    value = { currentTrack: null, isPlaying: false },
    props
  ) =>
    renderer.create(
      <PlayerContext.Provider value={value}>
        <TrackList {...requiredProps} {...mockActions} {...props} />
      </PlayerContext.Provider>
    );

  describe('@render', () => {
    it('default', () => {
      const tree = renderComponent();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('@actions', () => {
    beforeEach(() => {
      jest.restoreAllMocks();
    });
    describe('onClick', () => {
      it('click on track calls onSelect with right id', () => {
        const component = renderComponent();
        component.root
          .findAllByProps({ className: 'track' })[0]
          .props.onClick();

        expect(mockActions.onSelect).toHaveBeenCalledWith(
          requiredProps.songs[0]
        );
      });
    });
  });
});
