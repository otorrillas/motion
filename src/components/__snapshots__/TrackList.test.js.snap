// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrackList /> @render default 1`] = `
<table
  className="trackList"
>
  <thead>
    <tr>
      <th />
      <th>
        Song
      </th>
      <th>
        Artist
      </th>
      <th>
        Album
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      className="track"
      onClick={[Function]}
    >
      <td>
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-play fa-w-14 "
          data-icon="play"
          data-prefix="fas"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </td>
      <td>
        sample title
      </td>
      <td>
        sample artist
      </td>
      <td>
        sample album
      </td>
    </tr>
    <tr
      className="track"
      onClick={[Function]}
    >
      <td>
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-play fa-w-14 "
          data-icon="play"
          data-prefix="fas"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </td>
      <td>
        sample title 2
      </td>
      <td>
        sample artist 2
      </td>
      <td>
        sample album 2
      </td>
    </tr>
  </tbody>
</table>
`;
